[GTK] Port SelectionData to new IPC serialization format

csaavedra opened the pull request:
#### 41e8f107db1c092980668ad788326ae5b5354dd9
<pre>
[GTK] Port SelectionData to new IPC serialization format
<a href="https://bugs.webkit.org/show_bug.cgi?id=271836">https://bugs.webkit.org/show_bug.cgi?id=271836</a>

Reviewed by Carlos Garcia Campos.

Remove the ArgumentCoders for SelectionData, adapt the
class a bit to make it work well with the generated serializers,
and add a new ArgumentCodersGtk.serialization.in for the
GTK-specific serializers.

* Source/WebCore/platform/gtk/PasteboardGtk.cpp:
(WebCore::Pasteboard::typesSafeForBindings):
(WebCore::Pasteboard::readOrigin):
(WebCore::Pasteboard::readStringInCustomData):
* Source/WebCore/platform/gtk/SelectionData.cpp:
(WebCore::SelectionData::SelectionData):
* Source/WebCore/platform/gtk/SelectionData.h:
(WebCore::SelectionData::setImage):
(WebCore::SelectionData::image const):
(WebCore::SelectionData::customData const):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.cpp:
(IPC::encodeImage): Deleted.
(IPC::decodeImage): Deleted.
(IPC::ArgumentCoder&lt;SelectionData&gt;::encode): Deleted.
(IPC::ArgumentCoder&lt;SelectionData&gt;::decode): Deleted.
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.h:
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in: Added.
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::DragSource):
* Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
(WebKit::Clipboard::write):

Canonical link: <a href="https://commits.webkit.org/276991@main">https://commits.webkit.org/276991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c590209a5801a4bde3fa4b2abecfffb8eeb59533

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/46386 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/25527 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/48980 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/49061 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/42427 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/48693 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/29886 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/22983 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/37854 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/46964 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/22561 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/39993 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/19094 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/19932 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/41098 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/4431 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/42666 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/41462 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/50887 "Built successfully") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/21391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/17830 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/45089 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/22682 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/44016 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/10261 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/23052 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/22381 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebCore/platform/gtk/SelectionData.h:
```diff
@@ -66,13 +66,16 @@ class SelectionData {
 
     void setCustomData(Ref<SharedBuffer>&& buffer) { m_customData = WTFMove(buffer); }
-    SharedBuffer* customData() const { return m_customData.get(); }
+    RefPtr<SharedBuffer> customData() const { return m_customData; }
```
Why not `RefPtr<SharedBuffer>&` or `const RefPtr<SharedBuffer>&`

csaavedra replied:
Thanks!

carlosgcampos commented on Source/WebCore/platform/gtk/SelectionData.h:
```diff
@@ -53,7 +53,7 @@ class SelectionData {
 
     void setImage(Image* newImage) { m_image = newImage; }
-    Image* image() const { return m_image.get(); }
+    const RefPtr<Image>& image() const { return m_image; }
```
Could we change the setter to receive a `RefPtr<Image>&&`?

carlosgcampos commented on Source/WebCore/platform/gtk/PasteboardGtk.cpp:
```diff
@@ -350,7 +350,7 @@ Vector<String> Pasteboard::typesSafeForBindings(const String& origin)
     if (m_selectionData) {
         ListHashSet<String> types;
-        if (auto* buffer = m_selectionData->customData()) {
+        if (auto* buffer = m_selectionData->customData().get()) {
```
Wny don't we change `auto*` -> `auto&` and use the `RefPtr` instead of getting the pointer?

csaavedra replied:
True!

carlosgcampos commented on Source/WebCore/platform/gtk/SelectionData.h:
```diff
@@ -52,8 +52,8 @@ class SelectionData {
 
-    void setImage(Image* newImage) { m_image = newImage; }
-    Image* image() const { return m_image.get(); }
+    void setImage(RefPtr<Image>&& newImage) { m_image = newImage; }
```
```suggestion
    void setImage(RefPtr<Image>&& newImage) { m_image = WTFMove(newImage); }
```

carlosgcampos commented on Source/WebCore/platform/gtk/SelectionData.cpp:
```diff
@@ -26,6 +26,23 @@
 
 namespace WebCore {
 
+SelectionData::SelectionData(const String& text, const String& markup, const URL& url, const String& uriList, RefPtr<WebCore::Image> image, RefPtr<WebCore::SharedBuffer>&& buffer, bool canSmartReplace)
```
```suggestion
SelectionData::SelectionData(const String& text, const String& markup, const URL& url, const String& uriList, RefPtr<WebCore::Image>&& image, RefPtr<WebCore::SharedBuffer>&& buffer, bool canSmartReplace)
```

carlosgcampos commented on Source/WebCore/platform/gtk/SelectionData.cpp:
```diff
@@ -26,6 +26,23 @@
+    if (!uriList.isEmpty())
+        setURIList(uriList);
+    if (image)
+        setImage(image.get());
```
```suggestion
        setImage(WTFMove(image));
```

csaavedra committed:
[GTK] Port SelectionData to new IPC serialization format
https://bugs.webkit.org/show_bug.cgi?id=271836

Reviewed by Carlos Garcia Campos.

Remove the ArgumentCoders for SelectionData, adapt the
class a bit to make it work well with the generated serializers,
and add a new ArgumentCodersGtk.serialization.in for the
GTK-specific serializers.

* Source/WebCore/platform/gtk/PasteboardGtk.cpp:
(WebCore::Pasteboard::typesSafeForBindings):
(WebCore::Pasteboard::readOrigin):
(WebCore::Pasteboard::readStringInCustomData):
* Source/WebCore/platform/gtk/SelectionData.cpp:
(WebCore::SelectionData::SelectionData):
* Source/WebCore/platform/gtk/SelectionData.h:
(WebCore::SelectionData::setImage):
(WebCore::SelectionData::image const):
(WebCore::SelectionData::customData const):
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.cpp:
(IPC::encodeImage): Deleted.
(IPC::decodeImage): Deleted.
(IPC::ArgumentCoder<SelectionData>::encode): Deleted.
(IPC::ArgumentCoder<SelectionData>::decode): Deleted.
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.h:
* Source/WebKit/Shared/gtk/ArgumentCodersGtk.serialization.in: Added.
* Source/WebKit/UIProcess/API/gtk/DragSourceGtk3.cpp:
(WebKit::DragSource::DragSource):
* Source/WebKit/UIProcess/gtk/ClipboardGtk3.cpp:
(WebKit::Clipboard::write):

Canonical link: https://commits.webkit.org/276991@main
